sway/tree/root.c

```diff
@@ -0,0 +1,147 @@
+#define _POSIX_C_SOURCE 200809L
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+#include <wlr/types/wlr_output_layout.h>
+#include "sway/input/seat.h"
+#include "sway/tree/arrange.h"
+#include "sway/tree/container.h"
+#include "sway/tree/root.h"
+#include "sway/tree/workspace.h"
+#include "list.h"
+#include "log.h"
+
+struct sway_container root_container;
+
+static void output_layout_handle_change(struct wl_listener *listener,
+		void *data) {
+	arrange_windows(&root_container);
+	transaction_commit_dirty();
+}
+
+void root_create(void) {
+	root_container.id = 0; // normally assigned in new_swayc()
+	root_container.type = C_ROOT;
+	root_container.layout = L_NONE;
+	root_container.name = strdup("root");
+	root_container.instructions = create_list();
+	root_container.children = create_list();
+	root_container.current.children = create_list();
+	wl_signal_init(&root_container.events.destroy);
+
+	root_container.sway_root = calloc(1, sizeof(*root_container.sway_root));
+	root_container.sway_root->output_layout = wlr_output_layout_create();
+	wl_list_init(&root_container.sway_root->outputs);
+#ifdef HAVE_XWAYLAND
+	wl_list_init(&root_container.sway_root->xwayland_unmanaged);
+#endif
+	wl_list_init(&root_container.sway_root->drag_icons);
+	wl_signal_init(&root_container.sway_root->events.new_container);
+	root_container.sway_root->scratchpad = create_list();
+
+	root_container.sway_root->output_layout_change.notify =
+		output_layout_handle_change;
+	wl_signal_add(&root_container.sway_root->output_layout->events.change,
+		&root_container.sway_root->output_layout_change);
+}
+
+void root_destroy(void) {
+	// sway_root
+	wl_list_remove(&root_container.sway_root->output_layout_change.link);
+	list_free(root_container.sway_root->scratchpad);
+	//wlr_output_layout_destroy(root_container.sway_root->output_layout);
```

Comment:
If I uncomment this line then I reliably get this crash on shutdown (even with no views open):

    Thread 1 "sway" received signal SIGSEGV, Segmentation fault.
    0x00007ffff7ba75f6 in output_destroy (output=0x7ffff7ba7a53 <output_manager_send_details+58>) at ../subprojects/wlroots/types/wlr_xdg_output.c:76
    76		wl_resource_for_each_safe(resource, tmp, &output->resources) {
    (gdb) bt
    #0  0x00007ffff7ba75f6 in output_destroy (output=0x7ffff7ba7a53 <output_manager_send_details+58>) at ../subprojects/wlroots/types/wlr_xdg_output.c:76
    #1  0x00007ffff7ba7cde in wlr_xdg_output_manager_destroy (manager=0x555555976308) at ../subprojects/wlroots/types/wlr_xdg_output.c:242
    #2  0x00007ffff7ba7af8 in handle_layout_destroy (listener=0x555555976350, data=0x555555975de0) at ../subprojects/wlroots/types/wlr_xdg_output.c:203
    #3  0x00007ffff7baa13e in wlr_signal_emit_safe (signal=0x555555975e18, data=0x555555975de0) at ../subprojects/wlroots/util/signal.c:29
    #4  0x00007ffff7b97031 in wlr_output_layout_destroy (layout=0x555555975de0) at ../subprojects/wlroots/types/wlr_output_layout.c:66
    #5  0x0000555555597a6d in root_destroy () at ../sway/tree/root.c:52
    #6  0x000055555556096c in main (argc=4, argv=0x7fffffffe288) at ../sway/main.c:462

Any ideas?

Reply:
Don't destroy objects created with a `wl_display` argument - they destroy themselves automaticlly when the display goes down to prevent being used while the display is destroyed.

Reply:
Eh, `wlr_output_layout` doesn't have a `wl_display` argument. Then this is a bug in wlroots.